resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -368,14 +364,11 @@ private[spark] class KubernetesClusterSchedulerBackend(
    *         locality if an executor launches on the cluster node.
    */
   private def getNodesWithLocalTaskCounts() : Map[String, Int] = {
-    val executorPodsWithIPs = EXECUTOR_PODS_BY_IPS_LOCK.synchronized {
-      executorPodsByIPs.values.toList  // toList makes a defensive copy.
-    }
     val nodeToLocalTaskCount = mutable.Map[String, Int]() ++
       KubernetesClusterSchedulerBackend.this.synchronized {
         hostToLocalTaskCount
       }
-    for (pod <- executorPodsWithIPs) {
+    for ((_, pod) <- executorPodsByIPs) {
```

Comment:
My change is based on the understanding that we don't need the whole map to be consistent during the iteration the Map. The previous lock there was for thread-safety not for exclusive lock down of the map for access.
 "iterators are designed to be used by only one thread at a time." ==> means a single iterator cannot be shared across multiple threads safely. Multiple threads can safely iterate as long as each thread has its own copy of an iterator.